lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -405,6 +405,308 @@ bool ADContext::supportsFloatingPointDifferentiation(Type type) const {
   return fpConf.hasValue();
 }
 
+//===----------------------------------------------------------------------===//
+// Control flow canonicalization
+//===----------------------------------------------------------------------===//
+
+namespace {
+class ControlFlowCanonicalization {
+private:
+  SILFunction &function;
+  SILBuilder builder = SILBuilder(function);
+  DominanceInfo &domInfo;
+  SILLoopInfo &loopInfo;
+
+public:
+  explicit ControlFlowCanonicalization(SILFunction &function,
+                                       DominanceInfo &domInfo,
+                                       SILLoopInfo &loopInfo)
+    : function(function), domInfo(domInfo), loopInfo(loopInfo) {}
+
+  /// Run control flow canonicalization on the function.
+  bool run();
+};
+}
+
+bool ControlFlowCanonicalization::run() {
+  DEBUG(getDebugStream() << "Running control flow canonicalization on function "
+        << function.getName() << '\n');
+  bool changed = false;
+  assert(!function.isNoReturnFunction() && !function.isExternalDeclaration());
+  assert(function.findReturnBB().getNodePtr());
+  // Canonicalize loops.
+  canonicalizeAllLoops(&domInfo, &loopInfo);
+  // TODO: Handle multiple loop exits.
+  return changed;
+}
+
+//===----------------------------------------------------------------------===//
+// Activity Analysis
+//===----------------------------------------------------------------------===//
+
+/// In many real situations, the end-users of AD need only the derivatives of
+/// some selected outputs of `P` with respect to some selected inputs of `P`.
+/// Whatever the differentiation mode (tangent, reverse,...), these restrictions
+/// allow the AD tool to produce a much more efficient differentiated program.
+/// Essentially, fixing some inputs and neglecting some outputs allows AD to
+/// just forget about several intermediate differentiated variables.
+///
+/// Activity analysis is the specific analysis that detects these situations,
+/// therefore allowing for a better differentiated code. Activity analysis is
+/// present in all transformation-based AD tools.
+///
+/// To begin with, the end-user specifies that only some output variables (the
+/// “dependent”) must be differentiated with respect to only some input
+/// variables (the “independent”). We say that variable `y` depends on `x` when
+/// the derivative of `y` with respect to `x` is not trivially null. We say that
+/// a variable is “varied” if it depends on at least one independent. Conversely
+/// we say that a variable is “useful” if at least one dependent depends on it.
+/// Finally, we say that a variable is “active” if it is at the same time varied
+/// and useful. In the special case of the tangent mode, it is easy to check
```

Comment:
Perhaps start a new paragraph with `In the special case of the tangent mode...`

Reply:
I copied from https://www-sop.inria.fr/tropics/papers/supportCoursDA.pdf so I think better to preserve that for now.

Reply:
That's fair. The definitions of "depends on", "varied", "useful", and "active" are important and are kinda buried in this huge paragraph though.